README.md
# Who is on?

プログラミングの勉強をもう一年したい学生のための春休みお手本プロジェクト。

課題は研究室出場記録を取る Racket プログラム。
具体的には、
WiFi機器（ケータイ電話を想定している）を持った誰がいつ LAN に接続したかを記録し、
ウェブで表示する。
who-is-on を実行する PC と ケータイなどの WiFi 機器が同一 LAN に接続していることを前提とする。

* who-is-on-update.rkt は、ARP テーブルに見つかる MAC アドレスをデータベースに記録する。

* who-is-on-app.rkt は指定したユーザの MAC アドレスが記録された日時を求めに応じて表示する。

* update.sh は who-is-on-update.rkt を呼んだ後、
update.sh 自身の呼び出しをat でスケジュールする。

### description

* 現在、研究室に来ているユーザ、電源が入っている PC を
<span style='color:red;'>●</span>(赤丸)
で表示。(図はバージョン 0.13 のものです)

    ![現在状況](images/current.png)

* 研究室メンバーの過去の出席状況一覧。

    ![一覧](images/members.png)

* 数時間ごとの各メンバーの記録。

    ![詳細](images/personal.png)

### requirement

開発は macos, linuxmint 19.

* racket, sqlite3

```sh
$ sudo apt install racket sqlite3
```
* dmac/spin パッケージ

```sh
$ raco pkg install https://github.com/dmac/spin.git
```

### install

```sh
$ make create
$ cp env-template /some/where/.env
$ make install
$ racket ./who-is-on-update.rkt
$ racket ./who-is-on-app.sh
```

```
$ open http://localhost:8000/users
```

update.sh は who-is-on-update.rkt をほぼ１時間おきに実行する。
who-is-on.service はその systemd ファイル。

### FIXME/TODO

* 0.19 auth を独立した関数・モジュールに。

* 0.20 let で sql 発行するのは宮川に良くない。リファクタしよう。

* 必要ない MAC は集めない。

* なんだ、この uptime は？Racket のバージョンの違いか？

```
ubuntu@vm2019:/srv/who-is-on$ uptime
 13:10:01 up 6 days,  1:44,  2 users,  load average: 21.10, 15.35, 10.84
```

* who-is-on-update.rb に --verbose オプション

* when terminate who-is-on-app by ^C, visited end points will be echoed back.
  useless. danger. how to stop it?

* even if tmint pings to tmint itself, tmint's mac-addr does not
  appear on arp table.

```sh
$ ping 192.168.0.11
PING 192.168.0.11 (192.168.0.11) 56(84) bytes of data.
64 bytes from 192.168.0.11: icmp_seq=1 ttl=64 time=0.022 ms
64 bytes from 192.168.0.11: icmp_seq=2 ttl=64 time=0.036 ms
64 bytes from 192.168.0.11: icmp_seq=3 ttl=64 time=0.035 ms
^C

$ arp -an
? (192.168.0.14) at 00:11:32:9f:55:02 [ether] on enp0s5
? (192.168.0.32) at aa:b8:64:3e:d8:25 [ether] on enp0s5
? (192.168.0.1) at b0:ac:fa:cc:bf:d7 [ether] on enp0s5
? (192.168.0.2) at 68:5b:35:a0:53:74 [ether] on enp0s5
? (192.168.0.3) at c0:a5:3e:50:04:ee [ether] on enp0s5
$
```
* macos の at コマンド

    at: pluralization is wrong
    at: cannot open lockfile /usr/lib/cron/jobs/.lockfile: Operation not permitted

* nginx リバースプロキシーの設定方法

    名前ベースの仮想ホストは C104 での運用には適当ではない。
    パスベースでlocalhost:8000 へ振るんだが、

    * localhost の名前が使えないホストがある。
    * プロキシがつながらなくなる。sites-enable からのリンクでやった場合。

### FIXED/UPDATED

* 0.18 /users に本日スタート時間を表示。

* 0.17.5 amend. 出席取り損なった学生に。

* 0.17.3 /user/:name, 土日に色つける。

* 0.13 テーブルに cat コラム。並びを cat 順にする。

* 0.12 テーブルに jname。表示を日本語に。call/cc 使ったがまだダメだね。会得が不十分。

* 0.11 find-me.rkt。arp テーブルに乗らないユーザがある。

* 0.10 絵文字、README.md に図で説明。

* 0.9 /list で一覧表示。

* 0.8 別 LAN の hkimura 状況をプッシュ。

* 0.7.2 nginx で basic auth
  working tree の中で sed しない。

* 0.7 /users/ のページに現在値を表示

* app installer、url の書き換え、DB を上書きyes・noオプション

    →  make install で。

* query-exec の回数を減らす。

    → 1時間に一度実行するくらいの頻度で呼ぶ関数。血眼にならないでよい。

* app installer、url の書き換え、DB を上書きyes・noオプション

* install の sed ができない。

  →  エスケープじゃなく、セパレータを換える作戦で。

* 2019-03-14 10 分おきに cron から起こすとして、確率 1/3 で実行するのは？

  → アラウンド 60 分後に実行するにしよう。0.5.4.

* 2019-03-14 macos の /usr/sbin/arp では 00 を 0 に短縮して表示する。
  string= で比較できない。mac= を定義するとしても、SQLite3 に組み込むのは面倒だ。

    → padzero を定義。0.5.3.3

### MAC アドレスの取得

ICMP ブロードキャストに反応しないホストもあるため、
サブネットの有効なアドレス 10.0.33.1..254 の一つ一つにタイムアウト付き ping を打つ。

ping を直列に実行しては
（タイムアウト時間） x （サブネットの数）
だけ時間がかかってしまう。

ping を並列に実行する関数を Racket でどう定義するかが問題。スレッド使え。

### 定期実行

プログラム自体でゆっくりループするか、cron で、
と思ったが、at コマンド で自分自身を 55〜65 分後に呼び出す方法に変更。0.5.4

cron, at は linux/unix の基本的機能の一つ。

### 取得した Mac アドレスの記録

SQLite3 にタイムスタンプと共に記録する。

* SQL できるようになれ。

* Racket でデータベースを扱う具体的な方法を身につけろ。

### 出場記録の表示

Racket の web フレームワーク dmac/spin で web アプリを作成する。

* get /users

  ユーザ一覧の表示とユーザ名から記録へのリンク。
  １時間前までに記録があれば bullet を赤で。0.7

* get /user/name

  ユーザ name の記録を表示。

* get /user/name/yyyy-mm-dd

  ユーザ name の yyyy-mm-dd の記録。 /user/name があれば不要か。

* get /users/new, post /users/create

  ユーザを追加する。

ユーザの削除は 0.6.2 まで定義していない。

### Web アプリのデプロイ

流行りは nginx のリバースプロキシだろう。

---
hiroshi.kimura.0331@gmail.com


ChangeLog
2019-05-31  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.18 start-time
	* 0.17.5.2
	* who-is-on-app.rkt: BUGFIX: password 属性を pass で書いていた。

2019-05-30  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* who-is-on-app.rkt: ryuto-circuit へのリンク

2019-05-21  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.17.1 refactoring
	get /stays/:user under construction

2019-05-20  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.17 stay-time

2019-04-17  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.15.4 bugfix
	「いるよ」でエンバグ。c104 でオンサイトデバッグした内容をきちんと取り込んでいなかったか。
	別アルゴリズムで修正。

	* 0.15.1
	* in? タイプチェック。
	* x-real-ip: bytes->string の前にタイプチェック。

	* 0.15
	* who-is-on-app.rkt: レイアウトを現代風に。
	redmine と L99 へのリンクをナビに。

2019-04-12  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.14.2
	* who-is-on-app.rkt: (x-real-ip) ninx のプロキシーを通すわけで、
	アプリに届く client-ip は 127.0.0.1 になってしまう。
	nginx が付与する x-real-ip ヘッダから元の IP を拾う。

	* しかし、ヘッダを拾ってみると通信が IPv6 していることがわかった。
	arp6 を考えるのは面倒なので、ネットで探してサーバレベルで IPv6 を
	ディスエーブルしたところ、なんかおかしい。
	nginx が IPv6 をリッスンしてるよ！
	コンフィグを書き換えて nginx が IPv4 のみをリッスンするようにし、
	解決。細かいメッセージなどの調整はこの後。
	lxd のレベルで IPv6 止めても良かったかも。

2019-04-11  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.14 いるよボタン
	* who-is-on-app.rkt: (get "/i-m-here") とそれに関連する関数。
	* in? 同一サブネットか？
	* ip->mac arp テーブルを引いて ip に対応する mac アドレスを見つける。
	* wifi? mac は users テーブルに登録があるか？
	* arp 関連を who-is-on-update.rkt から分離独立させて arp.rkt とする。

2019-04-10  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.13
	* 並びを年の順に。

	* 0.12
	* who-is-on-app.rkt: *name-jname* 辞書
	* call/cc 初利用！(j name)
	* update (get "/list")
	* update (get "/users")
	* get/post users に jname 分追加
	* only-people: users から変更。関数の内容を表す名前に。

2019-04-09  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.10.1
	* update Makefile: install weekday.rkt

	* 0.10
	* merge miyakwa's weekday.rkt

2019-04-03  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.9.5
	あれ？0.9.4 飛ばしちゃったか？
	* who-is-on-app.rkt: gregor 見つけた。sunday? saturday? 利用して
	weekday? を定義。

	* 0.9.3 vm2019 で git pull しても、新しくならない。どうして？
	make install で VERSION をコピーするようにしよう。

	* 0.9.2 hotfix が期待通りに実行できなかった模様。新たに 0.9.2 リリース。

	* 0.9.1 update Makefile, who-is-on-app.rkt
	* グローバル変数 *users* の除去

	* 0.9 /list
	* who-is-on-app.rkt: wifis, dates-all.

2019-03-28  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.8
	* who-is-on-app.rkt: define funtion 'now'
	status? use the 'now' function

	* Makefile (production): working tree で sed しない。
	basic auth

2019-03-25  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.7.1
	who-is-on-app.rkt: asynchronous update.

2019-03-23  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.7 who-is-on-app.rkt: status? を定義。
	過去１時間以内に記録が見つかる場合、/users に赤で表示。

2019-03-18  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* update README.md
	* remove useless conf/crontab
	* who-is-on.service:EnvironmentFile must be /srv/who-is-on/.env

2019-03-17  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.6.1 Makefile (install-systemd): systemd のユーザからのインストール先は
	/lib/systemd/system でした。

	* 0.6 who-is-on-app.rkt: get "/"
	* env-template にマジ設定の一部を
	* .gitignore に who-is-on.sqlite3, .env, update.log を無視する


2019-03-16  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.5.5
	* Makefile (install)
	* Makefile (production, development): production 用に url を書き換える。

2019-03-14  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.5.4 update.sh: cron ではなく、コマンド at に 55 \pm 10 で
	自分自身を呼び出すスクリプト。
	* env/.placeholder: 動作に必要な環境変数を .placeholder に書いておく。
	* who-is-on-app.rkt: 127.0.0.1 しかリッスンしない。
	;;(run #:listen-ip #f #:port 8000)
	(run #:port 8000)
	* who-is-on-update.rkt: (pad mac) mac アドレス中のゼロを補完する。

	* 0.5.3.2 bugfix
	/bin/sh には source コマンドがない。bash にはある。
	fix typo, who-is-in → who-is-on
	* 0.5.3 /users/new, /users/create
	update.sh スクリプト。
	who-is-on-app.rkt: ページにバージョン表示。
	update Makefile

	* [0.5.4] cron で決め打ちではなく、
	プラスマイナス n 分 で Mac アドレス採集する。

	* [0.5.3] ユーザの追加ができる。

	* [0.5.3] 括弧つきのバージョン番号は「これから」の意味に使おう。
	予定って意味。

2019-03-13  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* [0.5] （外から curlで）動作確認。http://vm2019.local/who/users など。

	* conf/: crontab -l と nginx/site-enable/who-is-in の中身。

	* [0.4.1] update README.md

	* who-is-on-app.rkt: rewrite get /users use with-output-to-string.

	* who-is-on-update.rkt: rename from who-is-on.rkt

	* who-is-on-seed.rkt: rename from seed.rkt

	* who-is-on-app.rkt: rename from app.rkt

	* who-is-on.rkt: arp エントリーが incomplete な時はインサートしない。

	* [0.3.1] Makefile (git remote set-url origin ssh):
	リモートリポジトリを切り替える ds218j と syno2

	* mac addr ではなく、user.id を記録するようにするか。
	有効な user.id だけを記録する方がいいか。
	user.name でも。DB容量的には id だが。
	pings? も復活させる。


who-is-on-app.rkt
#!/usr/bin/env racket
#lang racket
;;;
;;; WiFi で出場状況チェック by hkimura
;;; make install will rewrite this file suitable for production
;;;
;;; update 2019-03-13,
;;;        2019-03-14,
;;;        2019-03-17,
;;;        2019-03-23,
;;;        2019-03-25 asynchronous update
;;;        2019-03-28 cancel 2019-03-25, define 'on'
;;;        2019-04-03 for*/list
;;;        2019-04-09 merge miyakawa's weekday.rkt
;;;        2019-04-10 japase name, display order
;;;        2019-04-11 いるよボタン
;;;        2019-04-17 レイアウト変更、redmine、l99 へのリンク、絵文字
;;;        2019-05-01 /list and /list-all
;;;        2019-05-20 滞留時間
;;;        2019-05-27 /list にコメント
;;;        2019-05-27 土日をカラー表示
;;;        2019-05-29 ryuto-circuit へのリンク
;;;        2019-05-30 amend, under construction
;;;        2019-05-31 start time
;;;        2019-06-04 as -- link to miyakawa assign.

(require db
         web-server/http
         (planet dmac/spin)
         "weekday.rkt"
         "arp.rkt")

(define VERSION "0.18.1")

(define sql3 (sqlite3-connect #:database (or (getenv "WIO_DB") "who-is-on.sqlite3")))

(define header
  "<!doctype html>
<html>
<head>
<meta name='viewport' content='width=device-width, initial-scale=1.0'/>
<link
 rel='stylesheet'
 href='https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css' integrity='sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T' crossorigin='anonymous'/>
<style>
.red   { color: red; }
.black { color: black; }
.blue  { color: blue; }
</style>
</head>
<body>
<div class='container'>
<h2>Who is on?</h2>
<p>
<a href='/users' class='btn btn-outline-primary btn-sm'>back</a>
<a href='https://rm4.melt.kyutech.ac.jp' class='btn btn-outline-primary btn-sm'>redmine</a>
<a href='https://l99.melt.kyutech.ac.jp' class='btn btn-outline-primary btn-sm'>L99</a>
<a href='http://rc.melt.kyutech.ac.jp:3000' class='btn btn-outline-primary btn-sm'>rc</a>
<a href='http://c104.melt.kyutech.ac.jp:8004' class='btn btn-outline-primary btn-sm'>as</a>
</p>")

(define footer
  (format "<hr>
hiroshi . kimura . 0331 @ gmail . com, ~a,
<a href='https://github.com/hkim0331/who-is-on'>github</a>
</div>
</body>
</html>
" VERSION))

(define (hh:mm:ss->sec s)
  (let ((ret (map string->number (string-split s ":"))))
    (+ (* (first ret) 3600) (* (second ret) 60) (third ret))))

(define (time-diff t2 t1)
  (- (hh:mm:ss->sec t2)
     (hh:mm:ss->sec t1)))

(define (stays user)
  (map
   vector->list
   (query-rows
    sql3
    "select date,time from mac_addrs
inner join users on users.wifi=mac_addrs.mac
where users.name=$1" user)))

(define (total-stay-second u)
  (define (first-last xs)
    (list (first xs) (first (reverse xs))))
  (define (diff xs)
    (time-diff (second (second xs))
               (second (first xs))))
  (let ((stays (stays u)))
    (apply + (map diff (map first-last (group-by first stays))))))

(define (users)
  (query-list sql3 "select name from users order by cat desc,name"))

(define (only-people)
  (filter (lambda (s) (not (regexp-match #rx"^imac" s))) (users)))

(define (now)
  (string-split (query-value sql3 "select datetime('now','localtime')")))

(define (dd-mm s)
  (substring s 5 10))

(define (weekdays days)
  (filter weekday? days))

;; macro?
(define (html . body)
  (format "~a~a~a"
    header
    (string-join body)
    footer))

;;want rewrite
(define (status? name)
  (define (hh s)
    (string->number (first (string-split s ":"))))
  (let*
      ((ret
        (query-rows
         sql3
         "select date,time from mac_addrs where mac=$1 order by id desc limit 1"
         (wifi name))))
    (if (null? ret)
        false
        (let* ((now (now)))
          (and (string=? (first now) (vector-ref (first ret) 0))
               ;; max 2 hours
               (<= (- (hh (second now)) (hh (vector-ref (first ret) 1))) 1))))))

;; CHECK: query-maybe-value?
(define (wifi name)
  (query-value sql3 "select wifi from users where name=$1" name))

(define (wifi? mac)
  (query-maybe-value sql3 "select id wifi from users where wifi=$1" mac))

(define (hh:mm s)
  (let ((ret (string-split s ":")))
    (format "~a:~a" (first ret) (second ret))))

(define (interpose sep xs)
  (define (I sep xs ret)
    (if (null? xs)
        ret
        (I sep (cdr xs) (cons sep (cons (car xs) ret)))))
  (if (null? xs)
      '()
      (reverse(cdr (I sep xs '())))))

(define (string-join-with sep strings)
  (string-join (interpose sep strings)))

;;FIXME, not DRY
(define (users-wifi)
  (let ((q (string-join-with
            "or"
            (map (lambda (s) (format "name='~a'" s)) (only-people)))))
    (query-list
     sql3
     (format "select wifi from users where ~a order by cat desc, name" q))))

(define (dates days)
  (query-list
   sql3
   (format "select distinct(date) from mac_addrs order by date desc limit ~a" days)))

(define (wifi->name wifi)
  (first (query-list sql3 "select name from users where wifi=$1" wifi)))

(define *name-jname*
  (query-rows sql3 "select name,jname from users"))

(define (insert mac)
  (let* ((now (now)))
    (query-exec
     sql3
     "insert into mac_addrs (mac,date,time) values ($1,$2,$3)"
     mac (first now) (second now))
    "OK"))

;; FIXME: if did not find name in *name-jname*, j must be return name.
(define (j name)
  (call/cc
   (lambda (return)
     (for ([pair *name-jname*])
       (when (string=? name (vector-ref pair 0))
         (return (vector-ref pair 1)))))))

;; not collect, but enough.
(define (in? ip)
  (let ((subnet (getenv "WIO_SUBNET")))
    (and subnet (regexp-match (regexp subnet) ip))))

(define (ip->mac ip)
  (call/cc
   (lambda (return)
     (for ([line (arp)])
       (when (string=?
              (first (regexp-match #rx"[0-9]+.[0-9]+.[0-9]+.[0-9]+" line))
              ip)
         (return (fourth (string-split line)))))
     (return #f))))

;;0.14.2
(define (x-real-ip req)
  (let ((headers
         (filter
          (lambda (x) (bytes=? #"X-Real-IP" (header-field x)))
          (request-headers/raw req))))
    (if (null? headers)
        "not found"
        (bytes->string/latin-1 (header-value (first headers))))))

;;;
;;; end points
;;;

;; 2019-05-30
;; 出席記録しそこなった学生のために
;; need auth
(get "/amend"
  (lambda (req)
(html
  "<h3>amend(under construction)</h3>
   <form method='post' action='/amend'>
    <p>admin <input name='admin'> password <input type='password' name='pass'></p>
    <p>who?<input name='name'></p>
    <p>when<input name='date' placeholder='yyyy-mm-dd'></p>
    <p><input type='submit' value='amend'></p>
</form>")))

(post "/amend"
  (lambda (req)
    (let ((admin (params req 'admin))
          (pass (params req 'pass))
          (wifi (wifi (params req 'name))))
      (if (string=? (params req 'pass)
                    (query-value sql3 "select pass from pass"))
        (begin
          (query-exec
            sql3
            "insert into mac_addrs (mac, date, time) values ($1, $2, $3)"
            wifi (params req 'date) "12:00:00")
          (html "<p>OK. <a href='/'>back</a></p>"))
        (html "<p>bad password. <a href='/amend'>amend</a> or <a href='/'>back</a></p>")))))

(get "/stays/:user"
     (lambda (req)
       (format "~a" (stays (params req 'user)))))

;; 本当は get じゃなくて post だな。
;; href='/i-m-here' で飛ばしたいがために get
(get "/i-m-here"
  (lambda (req)
    (let ((client-ip (x-real-ip req)))
      (if (in? client-ip)
        (let ((mac (pad (ip->mac client-ip))))
             (displayln (format "mac: ~a" mac))
          (if (wifi? mac)
            (begin
              (insert mac)
              (html (format "OK.")))
            (html (format "not registered."))))
        (html "not in C104.")))))

(get "/info"
  (lambda (req)
    (html
      (format "<p>WIO_DB: ~a</p>" (getenv "WIO_DB"))
      (format "<p>WIO_SUBNET: ~a</p>" (getenv "WIO_SUBNET")))))

;; http -f http://localhost:8000/on name=***** pass=*****
(post "/on"
      (lambda (req)
        (let ((pass (query-value sql3 "select pass from pass"))
              (wifi (query-maybe-value
                     sql3
                     "select wifi from users where name=$1"
                     (params req 'name))))
          (if (and (string=? pass (params req 'pass)) wifi)
              (insert wifi)
              "NG"))))

(get "/"
  (lambda ()
    (html
      "<p>try <a href='/users'>here</a>.</p>")))

(post "/users/create"
  (lambda (req)
    (query-exec
      sql3
      "insert into users (name, jname, wifi) values ($1, $2, $3)"
      (params req 'name)
      (params req 'jname)
      (params req 'wifi))
    (html "<p>OK.</p>")))

(get "/users/new"
  (lambda (req)
    (html
     (with-output-to-string
       (lambda ()
          (displayln "<form method='post' action='/users/create'>")
          (displayln "<table>")
          (displayln "<tr><th>name</th><td><input name='name'></td></tr>")
          (displayln "<tr><th>日本名</th><td><input name='jname'></td></tr>")
          (displayln "<tr><th>wifi</th><td><input name='wifi'></td></tr>")
          (displayln "</table>")
          (displayln "<p><input type='submit' class='btn btn-primary' value='add'></p>")
          (displayln "</form>"))))))

;;2019-05-31, 0.18
(define (start-time uname day)
  (query-maybe-value
   sql3
   "select time from mac_addrs where mac=$1 and date=$2 order by time limit 1"
   (wifi uname) day))

(get "/users"
  (lambda (req)
    (html
     (with-output-to-string
       (lambda ()
         (displayln "<p>( ) は 4 月から通算の研究室滞留時間。<br>
卒論、PBL、だいじょうぶ？</p>")
         (displayln "<div class='container'><table>")
         (for ([u (users)])
           (displayln
            (format "<tr><td>~a</a></td><td><a href='/user/~a'>~a</a> (~a~a)</td><tr>"
                    (if (status? u) "😀" "▪️")
                    u
                    (j u)
                    (quotient (total-stay-second u) 3600)
                    (let ((st (start-time u (first (now)))))
                      (if st
                          (format ", ~a~~" st)
                          "")))))
         (displayln "</table></div><br>")
         (displayln
          "<p>
<a href='/i-m-here' class='btn btn-outline-primary btn-sm'>いるよ 😀</a>
<a href='/list' class='btn btn-outline-primary btn-sm'>list</a>
<a href='/amend' class='btn btn-primary btn-sm'>amend</a>
<a href='/users/new' class='btn btn-primary btn-sm'>add</a>
</p>"))))))

(get "/user/:name/:date"
  (lambda (req)
    (let* ((name (params req 'name))
           (date (params req 'date))
           (ret (query-list sql3 "select time from mac_addrs inner join users on mac_addrs.mac=users.wifi where users.name=$1 and date=$2"
                 name date)))
      (html
        (format "<h3>~a, ~a</h3>" name date)
        "<p>"
        (string-join (map hh:mm ret) " &rarr; ")
        "</p>"))))

;;FIXME: hash or vector?
(define (color n)
  (cond
    ((= n 0) "red")
    ((= n 6) "blue")
    (else "black")))

;; color weekends
(get "/user/:name"
  (lambda (req)
    (define date first)
    (define (first-date x) (date (first x)))
    (let ((name (params req 'name)))
      (html
        (format "<h3>~a</h3>" name)
        (with-output-to-string
          (lambda ()
            (let loop ((ret (map vector->list (query-rows sql3 "select date,time from  mac_addrs inner join users on mac_addrs.mac=users.wifi where users.name=$1 order by date desc, time" name))))
              (unless (null? ret)
                (let ((the-date (first-date ret)))
                (display (format "<p><b class='~a'>~a</b> "
                  (color (day-of-week the-date))
                  (dd-mm the-date)))
                (display
                  (string-join
                    (map (lambda (x) (hh:mm (second x)))
                      (filter (lambda (s) (string=? (date s) the-date)) ret))
                    " &rarr; "))
                (display "</p>")
                (loop (filter (lambda (s) (string<? (date s) the-date)) ret)))))))))))

;; 2019-05-01
;; bug: must not redefine list!
(define (list-days days)
  (let ((users (users-wifi))
        (dates (weekdays (dates days)))
        (status (query-rows sql3 "select date,mac from mac_addrs group by date")))
       (with-output-to-string
        (lambda ()
          (display "<p>課題やってこない大学に来ないじゃ救いようない。</p>")
          (display "<table>")
          (display "<tr><th></th>")
          (for ([u users])
            (let ((name (wifi->name u)))
              (display (format "<td><a href='/user/~a'>~a</a>|</td>"
                               name
                               (j name)))))
          (display "</tr>")
          (for ([d dates])
            (let ((st
                   (query-list
                    sql3
                    "select distinct(mac) from mac_addrs where date=$1" d)))
              (display (format "<tr><th>~a</th>" (dd-mm d)))
              (for ([u users])
                (display (format "<td style='text-align:center;'>~a</td>"
                                 (if (member u st string=?)
                                     "😀"
                                     "")))))
            (display "</tr>"))
          (display "</table>")))))


(get "/list-all"
     (lambda (req)
       (html
        (list-days 400))))

(get "/list"
     (lambda (req)
       (html
        (list-days 30))))

;; for debug only
(define r #f)
(get "/info"
     (lambda (req)
      (set! r req)
      (html
        (format "<p>WIO_DB: ~a</p>" (getenv "WIO_DB"))
        (format "<p>WIO_SUBNET: ~a</p>" (getenv "WIO_SUBNET"))
        (format "<p>x-real-ip: ~a</p>" (x-real-ip req)))))

;;
;; start server
;;
(displayln "start at 8000/tcp")
(run #:listen-ip "127.0.0.1" #:port 8000)
;; for debug
;;(run #:listen-ip #f #:port 8000)
